src/common.rs

```diff
@@ -251,6 +251,60 @@ macro_rules! bytes_operation_impl {
 
 pub(crate) use bytes_operation_impl;
 
+/// Implements [`core::iter::Sum`] and [`core::iter::Product`] for an integer type.
+macro_rules! impl_sum_product {
+    ($type:ident, $one:literal) => {
+        // This implements `Sum` for owned values, for example when using an iterator from a fixed-sized array.
+        impl<T, const BITS: usize> core::iter::Sum for $type<T, BITS>
+        where
+            Self: Integer + Default + core::ops::Add<Output = Self>,
+        {
+            #[inline]
+            fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+                // Use `default()` to construct a value of zero.
+                iter.fold(Self::default(), |lhs, rhs| lhs + rhs)
+            }
+        }
+
+        // This implements `Sum` for borrowed values, for example when using an iterator from a slice.
+        impl<'a, T, const BITS: usize> core::iter::Sum<&'a Self> for $type<T, BITS>
+        where
+            Self: Integer + Default + core::ops::Add<Output = Self>,
+        {
+            #[inline]
+            fn sum<I: Iterator<Item = &'a Self>>(iter: I) -> Self {
+                iter.fold(Self::default(), |lhs, rhs| lhs + *rhs)
+            }
+        }
+
+        // We need to use `Integer::from_()` to construct a value of one,
+        // which isn't available with `const_convert_and_const_trait_impl`.
+        #[cfg(not(feature = "const_convert_and_const_trait_impl"))]
+        impl<T, const BITS: usize> core::iter::Product for $type<T, BITS>
+        where
+            Self: Integer + core::ops::Mul<Output = Self>,
+        {
+            #[inline]
+            fn product<I: Iterator<Item = Self>>(iter: I) -> Self {
+                iter.fold(Self::from_($one), |lhs, rhs| lhs * rhs)
+            }
+        }
+
+        #[cfg(not(feature = "const_convert_and_const_trait_impl"))]
+        impl<'a, T, const BITS: usize> core::iter::Product<&'a Self> for $type<T, BITS>
+        where
+            Self: Integer + core::ops::Mul<Output = Self>,
+        {
+            #[inline]
+            fn product<I: Iterator<Item = &'a Self>>(iter: I) -> Self {
+                iter.fold(Self::from_($one), |lhs, rhs| lhs * *rhs)
```

Comment:
This looks a bit awkward because operator overloading apparently doesn't auto-deref. This isn't an issue for primitives because the standard library explicitly implement the operator traits for references, I'll do the same here in another PR.